torch/fx/experimental/symbolic_shapes.py

```diff
@@ -1321,9 +1321,20 @@ def method_to_operator(method):
         op = getattr(operator, method_attr)
     return op
 
-def cast_symbool_to_symint_guardless(symbool: torch.SymBool) -> torch.SymInt:
-    int_sym = sympy.Piecewise((1, symbool.node.expr), (0, True))
-    return symbool.node.shape_env.create_symintnode(int_sym, hint=int(symbool.node.require_hint()))
+def cast_symbool_to_symint_guardless(maybe_symbool: Union[bool, SymBool]) -> torch.SymInt:
+    from torch.fx.experimental.proxy_tensor import thunkify, set_proxy_slot, ProxyTorchDispatchMode
+
+    if isinstance(maybe_symbool, bool):
+        return int(maybe_symbool)
+
+    int_sym = sympy.Piecewise((1, maybe_symbool.node.expr), (0, True))
+    out = maybe_symbool.node.shape_env.create_symintnode(int_sym, hint=int(maybe_symbool.node.require_hint()))
+    if current_mode := torch.utils._python_dispatch._get_current_dispatch_mode():
```

Comment:
Seems like you want some sort of higher order op here, maybe?

Reply:
I think higher order op might be a bit overshooting since this op doesn't have a function as input and don't need functionalization etc. 

I'd prefer to keep it as what it is right now but not entirely sure..

Reply:
The main problem is your dispatch logic here is wrong (what if proxy is in the stack, but it's not the current mode), and reimplementing it all the way correctly is essentially the same as HOP.